pinot-core/src/test/java/org/apache/pinot/core/plan/maker/MetadataAndDictionaryAggregationPlanMakerTest.java

```diff
@@ -117,8 +119,9 @@ public void buildSegment()
   public void loadSegment()
       throws Exception {
     _indexSegment = ImmutableSegmentLoader.load(new File(INDEX_DIR, SEGMENT_NAME), ReadMode.heap);
-    _upsertIndexSegment = ImmutableSegmentLoader
-        .load(new File(INDEX_DIR, SEGMENT_NAME), ReadMode.heap, new PartitionUpsertMetadataManager());
+    _upsertIndexSegment = ImmutableSegmentLoader.load(new File(INDEX_DIR, SEGMENT_NAME), ReadMode.heap);
+    ((ImmutableSegmentImpl) _upsertIndexSegment)
+        .enableUpsert(new PartitionUpsertMetadataManager(), new ThreadSafeMutableRoaringBitmap());
```

Comment:
nit: I think it's preferred to enable it as early as possible (i.e in the constructor), we know this segment will be an upsert one.

Reply:
I moved this method out of the constructor because the segment loader doesn't need to know whether the segment has upsert enabled or not, and upsert metadata is updated after the segment is loaded.
Decoupling upsert from the segment loader help simplify the code path for the loading part, and keep all the upsert handling at the same place: `RealtimeTableDataManager.addSegment(ImmutableSegment immutableSegment)`

Reply:
okay, that's fair